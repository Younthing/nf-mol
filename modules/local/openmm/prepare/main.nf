process OPENMM_PREPARE {
    // 1. 进程标识和资源设置
    tag "${meta.id}"
    label 'process_low'

    // 2. 错误处理策略
    errorStrategy 'retry'
    maxRetries 3

    // 3. 环境配置
    conda "${moduleDir}/environment.yml"

    input:
    tuple val(meta), path(pdb_file, stageAs: 'input.pdb')

    output:
    tuple val(meta), path("*.{pdb,cif,pdbx}"), emit: structure
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    // 解析格式参数并设置扩展名
    def format = 'pdb'
    if (args.contains('--format pdb') || args.contains('-f pdb')) {
        format = 'pdb'
    }
    else if (args.contains('--format cif') || args.contains('-f cif')) {
        format = 'cif'
    }
    else if (args.contains('--format pdbx') || args.contains('-f pdbx')) {
        format = 'pdbx'
    }

    """
    prepare_pdb.py prepare \\
        ${pdb_file} \\
        --output ${prefix}_prepared.${format} \\
        ${args}

    # 获取版本信息到变量中，避免直接输出失败
    OPENMM_VERSION=\$(python -c "import openmm; print(openmm.__version__)" || echo "unknown")
    PDBFIXER_VERSION=\$(python -c "import pdbfixer; print('1.11')" || echo "unknown")

    # 创建版本文件
    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        openmm: "\${OPENMM_VERSION}"
        pdbfixer: "\${PDBFIXER_VERSION}"
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    // 在存根模式下也需要解析正确的格式
    def format = 'pdb'
    if (task.ext.args?.contains('--format pdb') || task.ext.args?.contains('-f pdb')) {
        format = 'pdb'
    }
    else if (task.ext.args?.contains('--format cif') || task.ext.args?.contains('-f cif')) {
        format = 'cif'
    }
    else if (task.ext.args?.contains('--format pdbx') || task.ext.args?.contains('-f pdbx')) {
        format = 'pdbx'
    }

    """
    touch ${prefix}_prepared.${format}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        openmm: "7.7.0"
        pdbfixer: "1.9"
    END_VERSIONS
    """
}
