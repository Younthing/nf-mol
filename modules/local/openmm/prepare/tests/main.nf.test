// 定义nextflow进程测试框架
nextflow_process {
    // 定义测试进程的基本信息
    name "Test Process OPENMM_PREPARE"    // 测试名称
    script "../main.nf"                   // 要测试的脚本路径
    process "OPENMM_PREPARE"              // 要测试的进程名称

    // 测试标签，用于分类和过滤测试
    tag "modules"                         // 表示这是模块测试
    tag "modules_nfcore"                  // 表示这是nf-core模块
    tag "openmm"                         // OpenMM相关测试
    tag "openmm/prepare"                 // 特定到prepare功能测试

    // 测试用例1: 标准PDB结构准备测试
    test("PDB structure preparation") {
        when {
            // 定义测试输入
            process {
                """
                input[0] = [
                    [ id:'1fjs', single_end:false ],  // 元数据映射，包含结构ID
                    file("${projectDir}/data/openmm/6w70.pdb")  // 输入PDB文件路径
                ]
                """
            }
        }

        then {
            // 定义测试断言
            assertAll(
                { assert process.success },                    // 验证进程是否成功完成
                // { assert snapshot(process.out).match() },      // 验证输出是否与快照匹配
                { assert path("${process.out.structure[0][1]}").exists() },  // 验证结构文件是否存在
                { assert path("${process.out.versions[0]}").exists() },      // 验证版本文件是否存在
                { assert path("${process.out.structure[0][1]}").getText().contains("ATOM") },  // 验证输出文件包含ATOM记录
                { assert path("${process.out.versions[0]}").getText().contains("openmm") }     // 验证版本信息包含openmm
            )
        }
    }

    // 测试用例2: 存根模式测试
    test("PDB structure preparation - stub") {
        options "-stub"    // 启用存根模式，用于快速测试而不实际运行计算

        when {
            // 定义与标准测试相同的输入
            process {
                """
                input[0] = [
                    [ id:'1fjs', single_end:false ],  // 元数据映射
                    file("${projectDir}/data/openmm/1fjs_protein.pdb")  // 输入文件路径
                ]
                """
            }
        }

        then {
            // 存根模式下的简化断言
            assertAll(
                { assert process.success },                    // 验证进程是否成功完成
                { assert snapshot(process.out).match() },      // 验证输出是否与快照匹配
                { assert path("${process.out.structure[0][1]}").exists() },  // 验证是否创建了占位输出文件
                { assert path("${process.out.versions[0]}").exists() }       // 验证是否创建了版本文件
            )
        }
    }

}
