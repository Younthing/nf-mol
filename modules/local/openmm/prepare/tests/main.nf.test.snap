{
    "PDB structure preparation - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "1fjs",
                            "single_end": false
                        },
                        "1fjs_prepared.pdb:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,cfdb612eb3b5a1921ded23a8bc098777"
                ],
                "structure": [
                    [
                        {
                            "id": "1fjs",
                            "single_end": false
                        },
                        "1fjs_prepared.pdb:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,cfdb612eb3b5a1921ded23a8bc098777"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-02-21T16:05:31.823991534"
    },
    "PDB structure preparation": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "1fjs",
                            "single_end": false
                        },
                        "1fjs_prepared.pdb:md5,e2ddd4239ef09dcd5da687476f7d9006"
                    ]
                ],
                "1": [
                    "versions.yml:md5,0ecf7938dc9d54af9a62f03b8643593e"
                ],
                "structure": [
                    [
                        {
                            "id": "1fjs",
                            "single_end": false
                        },
                        "1fjs_prepared.pdb:md5,e2ddd4239ef09dcd5da687476f7d9006"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,0ecf7938dc9d54af9a62f03b8643593e"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-02-21T17:36:55.008842189"
    }
}